engine/consensus/compliance/engine.go

```diff
@@ -66,49 +54,113 @@ func NewEngine(
 	me module.Local,
 	prov network.Engine,
 	core *Core) (*Engine, error) {
-	e := &Engine{
-		unit:             engine.NewUnit(),
-		log:              log.With().Str("compliance", "engine").Logger(),
-		me:               me,
-		mempool:          core.mempool,
-		metrics:          core.metrics,
-		headers:          core.headers,
-		payloads:         core.payloads,
-		state:            core.state,
-		tracer:           core.tracer,
-		prov:             prov,
-		core:             core,
-		pendingEventSink: make(EventSink),
-		blockSink:        make(EventSink),
-		voteSink:         make(EventSink),
-	}
-
-	var err error
-	// register the core with the network layer and store the conduit
-	e.con, err = net.Register(engine.ConsensusCommittee, e)
-	if err != nil {
-		return nil, fmt.Errorf("could not register core: %w", err)
-	}
 
 	// FIFO queue for block proposals
-	e.pendingBlocks, err = fifoqueue.NewFifoQueue(
+	pendingBlocks, err := fifoqueue.NewFifoQueue(
 		fifoqueue.WithCapacity(defaultBlockQueueCapacity),
-		fifoqueue.WithLengthObserver(func(len int) { e.mempool.MempoolEntries(metrics.ResourceBlockProposalQueue, uint(len)) }),
+		fifoqueue.WithLengthObserver(func(len int) { core.mempool.MempoolEntries(metrics.ResourceBlockProposalQueue, uint(len)) }),
 	)
 	if err != nil {
 		return nil, fmt.Errorf("failed to create queue for inbound receipts: %w", err)
 	}
 
 	// FIFO queue for block votes
-	e.pendingVotes, err = fifoqueue.NewFifoQueue(
+	pendingVotes, err := fifoqueue.NewFifoQueue(
 		fifoqueue.WithCapacity(defaultVoteQueueCapacity),
-		fifoqueue.WithLengthObserver(func(len int) { e.mempool.MempoolEntries(metrics.ResourceBlockVoteQueue, uint(len)) }),
+		fifoqueue.WithLengthObserver(func(len int) { core.mempool.MempoolEntries(metrics.ResourceBlockVoteQueue, uint(len)) }),
 	)
 	if err != nil {
 		return nil, fmt.Errorf("failed to create queue for inbound approvals: %w", err)
 	}
 
-	return e, nil
+	// define message queueing behaviour
+	handler, notify := engine.NewMessageHandler(
+		log,
+		engine.Pattern{
+			Match: func(msg *engine.Message) bool {
+				switch msg.Payload.(type) {
```

Comment:
How about this approach?
```suggestion
				_, ok := msg.Payload.(*events.SyncedBlock)
				return ok
```